handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1447,65 +1484,224 @@ private static ByteBuffer toByteBuffer(ByteBuf out, int index, int len) {
                 out.nioBuffer(index, len);
     }
 
+    private static boolean inEventLoop(Executor executor) {
+        return executor instanceof EventExecutor && ((EventExecutor) executor).inEventLoop();
+    }
+
     /**
-     * Fetches all delegated tasks from the {@link SSLEngine} and runs them via the {@link #delegatedTaskExecutor}.
-     * If the {@link #delegatedTaskExecutor} is {@link ImmediateExecutor}, just call {@link Runnable#run()} directly
-     * instead of using {@link Executor#execute(Runnable)}.  Otherwise, run the tasks via
-     * the {@link #delegatedTaskExecutor} and wait until the tasks are finished.
+     * Will either run the delegating task directly calling {@link Runnable#run()} and return {@code true} or will
+     * offload the delegating task using {@link Executor#execute(Runnable)} and return {@code false}.
+     *
+     * If the task is offloaded it will take care to resume its work on the {@link EventExecutor} once there are no
+     * more tasks to process.
      */
-    private void runDelegatedTasks() {
-        if (delegatedTaskExecutor == ImmediateExecutor.INSTANCE) {
+    private boolean runDelegatedTasks(boolean inUnwrap) {
+        if (delegatedTaskExecutor == ImmediateExecutor.INSTANCE || inEventLoop(delegatedTaskExecutor)) {
```

Comment:
For simplicity/no special case/unify maybe omit the ImmediateExecutor.INSTANCE check and let inEventLoop() take care of it.

Reply:
You mean basically move all the code to the new method ?

Reply:
Oh, I think it's nvm. ImmediateExecutor is not an EventExecutor (mixed it up with ImmediateEventExecutor).